# Coding Exercise (Pandas)

---

In [ ]:
!pip install --upgrade gdown

  Attempting uninstall: gdown
    Found existing installation: gdown 4.7.3
    Uninstalling gdown-4.7.3:
      Successfully uninstalled gdown-4.7.3


Loading the dataset -

In [ ]:
!gdown 1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd
!gdown 1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm

Downloading...
From: https://drive.google.com/uc?id=1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd
To: /content/movies.csv
100% 112k/112k [00:00<00:00, 2.67MB/s]
Downloading...
From: https://drive.google.com/uc?id=1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm
To: /content/directors.csv
100% 65.4k/65.4k [00:00<00:00, 4.08MB/s]


In [ ]:
import pandas as pd
import numpy as np

movies = pd.read_csv('movies.csv', index_col=0)
directors = pd.read_csv('directors.csv', index_col=0)

In [ ]:
movies.head()

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
5                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
5         4767  2007   May   Tuesday

In [ ]:
directors.head()

director_name    id gender
0      James Cameron  4762   Male
1     Gore Verbinski  4763   Male
2         Sam Mendes  4764   Male
3  Christopher Nolan  4765   Male
4     Andrew Stanton  4766   Male

---

### Q1 - How many years of data is given to you in the `movies` dataset?

In [ ]:
no_of_years = movies["year"].max() - movies["year"].min()
print(no_of_years)

40


---

### Q2 - Display the `ReleaseDate` in the Day-Month-Year format.

In [ ]:
releaseDate = movies["day"] + "-" + movies["month"] + "-" + movies["year"]
print(releaseDate[:5])

TypeError: can only concatenate str (not "int") to str

In [ ]:
# confirm the error by checking the datatypes of the columns
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            1465 non-null   int64  
 1   budget        1465 non-null   int64  
 2   popularity    1465 non-null   int64  
 3   revenue       1465 non-null   int64  
 4   title         1465 non-null   object 
 5   vote_average  1465 non-null   float64
 6   vote_count    1465 non-null   int64  
 7   director_id   1465 non-null   int64  
 8   year          1465 non-null   int64  
 9   month         1465 non-null   object 
 10  day           1465 non-null   object 
dtypes: float64(1), int64(7), object(3)
memory usage: 126.0+ KB


As year is in `int` format, we need to typecast it before concatenation.

We'll use the `astype()` function here for typecasting.

In [ ]:
releaseDate = movies["day"] + "-" + movies["month"] + "-" + movies["year"].astype("str")
print(releaseDate[:5])

0    Thursday-Dec-2009
1    Saturday-May-2007
2      Monday-Oct-2015
3      Monday-Jul-2012
4     Tuesday-May-2007
dtype: object


---

### Q3 - What is the % of male directors in the given data?

In [ ]:
directors["gender"].value_counts()

Male      1574
Female     150
Name: gender, dtype: int64

In [ ]:
directors["gender"].value_counts(normalize=True)

Male      0.912993
Female    0.087007
Name: gender, dtype: float64

In [ ]:
directors["gender"].value_counts(normalize=True)*100

Male      91.299304
Female     8.700696
Name: gender, dtype: float64

More than 91% of the directors in the dataset are Male.

---

### Q4 - How many different movies have been directed in the year 2010?

In [ ]:
movies2010 = movies[movies["year"]==2010]

In [ ]:
len(movies2010) # as every row is a unique movie

63

In [ ]:
movies2010["id"].nunique()

63

---

### Q5 - How many different directors have directed a movie since 2010 (inclusive)?

In [ ]:
sample = movies[movies["year"] >= 2010]
sample["director_id"].nunique()

163

---

### Q6 - What are the names of those directors?

In [ ]:
ids = sample["director_id"].unique()
print(ids[:5])

[4764 4765 4771 4763 4775]


In [ ]:
names = directors[directors["id"].isin(ids)]["director_name"]
print(names)

1            Gore Verbinski
2                Sam Mendes
3         Christopher Nolan
5                 Sam Raimi
9               Zack Snyder
               ...         
1132          Brad Anderson
1189         Michael Polish
1208          Paul Schrader
1294    Darren Lynn Bousman
1442      Nicole Holofcener
Name: director_name, Length: 163, dtype: object


In [ ]:
names.count() # validating

163

---

### Q7 - Display the name of the director who directed the Spider-Man 3 movie.

In [ ]:
# Approach 1: without using merge

df = movies[movies["title"] == "Spider-Man 3"]
df

id     budget  popularity    revenue         title  vote_average  \
4  43602  258000000         115  890871626  Spider-Man 3           5.9   

   vote_count  director_id  year month      day  
4        3576         4767  2007   May  Tuesday

In [ ]:
director_id = df["director_id"]
director_id

4    4767
Name: director_id, dtype: int64

In [ ]:
director_id.reset_index(drop=True, inplace=True)
director_id

0    4767
Name: director_id, dtype: int64

In [ ]:
id = director_id[0]

In [ ]:
name = directors.loc[directors["id"] == id, "director_name" ]
print(name)

5    Sam Raimi
Name: director_name, dtype: object


In [ ]:
# Approach 2: first merge and then filter row

data = movies.merge(directors, how='inner', left_on='director_id',right_on='id')
data.drop(['director_id','id_y'], axis=1, inplace=True)

In [ ]:
data.head()

id_x     budget  popularity     revenue                       title  \
0  43597  237000000         150  2787965087                      Avatar   
1  43622  200000000         100  1845034188                     Titanic   
2  43876  100000000         101   520000000  Terminator 2: Judgment Day   
3  43879  115000000          38   378882411                   True Lies   
4  44184   70000000          24    90000098                   The Abyss   

   vote_average  vote_count  year month        day  director_name gender  
0           7.2       11800  2009   Dec   Thursday  James Cameron   Male  
1           7.5        7562  1997   Nov    Tuesday  James Cameron   Male  
2           7.7        4185  1991   Jul     Monday  James Cameron   Male  
3           6.8        1116  1994   Jul   Thursday  James Cameron   Male  
4           7.1         808  1989   Aug  Wednesday  James Cameron   Male

In [ ]:
name = data[data["title"] == "Spider-Man 3"]["director_name"]
print(name)

29    Sam Raimi
Name: director_name, dtype: object


---

### Q8 - How many different movies have been directed by James Cameron between year 2005 to 2010?

In [ ]:
directors.head()

director_name    id gender
0      James Cameron  4762   Male
1     Gore Verbinski  4763   Male
2         Sam Mendes  4764   Male
3  Christopher Nolan  4765   Male
4     Andrew Stanton  4766   Male

In [ ]:
james_id = directors.loc[directors['director_name'] == "James Cameron"]
james_id

director_name    id gender
0  James Cameron  4762   Male

In [ ]:
james_id = directors.loc[directors['director_name'] == "James Cameron"].reset_index(drop = True)["id"][0]
james_id

4762

In [ ]:
movies.head()

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

In [ ]:
james_movies = movies.loc[(movies["director_id"] == james_id) & (movies["year"] >= 2000) & (movies["year"] <= 2015)]
james_movies

id     budget  popularity     revenue   title  vote_average  vote_count  \
0  43597  237000000         150  2787965087  Avatar           7.2       11800   

   director_id  year month       day  
0         4762  2009   Dec  Thursday

Let us write a function to do the exact same thing for whichever director we want.

In [ ]:
def solve(director_name, start_year, end_year):

  director_id = directors.loc[ directors['director_name'] == director_name].reset_index(drop = True)["id"][0]

  director_movies = movies.loc[(movies["director_id"] == director_id)]

  df = director_movies.loc[ (director_movies["year"] >= start_year) & (director_movies["year"] <= end_year) ]

  return df

In [ ]:
solve("James Cameron", 2000, 2015)

id     budget  popularity     revenue   title  vote_average  vote_count  \
0  43597  237000000         150  2787965087  Avatar           7.2       11800   

   director_id  year month       day  
0         4762  2009   Dec  Thursday

In [ ]:
solve("Gore Verbinski", 2000, 2015)

id     budget  popularity     revenue  \
1     43598  300000000         139   961000000   
8     43609  200000000         145  1065659812   
9     43610  255000000          49    89289910   
107   43775  135000000          29   245724603   
119   43796  140000000         271   655011224   
633   44780   57000000          21   147845033   
1040  45801   20000000          14    12482775   

                                                  title  vote_average  \
1              Pirates of the Caribbean: At World's End           6.9   
8            Pirates of the Caribbean: Dead Man's Chest           7.0   
9                                       The Lone Ranger           5.9   
107                                               Rango           6.6   
119   Pirates of the Caribbean: The Curse of the Bla...           7.5   
633                                         The Mexican           5.8   
1040                                    The Weather Man           6.0   

      vote_count  director_id  year month        day  
1           4500         4763  2007   May   Saturday  
8           5246         4763  2006   Jun    Tuesday  
9           2311         4763  2013   Jul  Wednesday  
107         2051         4763  2011   Mar  Wednesday  
119         6985         4763  2003   Jul  Wednesday  
633          441         4763  2001   Mar   Thursday  
1040         289         4763  2005   Oct   Thursday

In [ ]:
solve("Aditya Jain", 2000, 2015)

KeyError: 0

We get an error whenever the passed director name is not present in the dataset.

From the industry's perspective, we try to make our function robust to such conditions.

In [ ]:
def solve(director_name, start_year, end_year):

  df = directors[directors["director_name"] == director_name]

  if len(df) != 0:
    director_id = directors.loc[ directors['director_name'] == director_name].reset_index(drop = True)["id"][0]
    director_movies = movies.loc[(movies["director_id"] == director_id)]
    df = director_movies.loc[ (director_movies["year"] >= start_year) & (director_movies["year"] <= end_year) ]
    return df

  else:
    return "No movies directed!"

In [ ]:
solve("Aditya Jain", 2000, 2015)

'No movies directed!'

---

### Q9 - Among the top 10 most popular movies, which is the movie with the 2nd highest vote_avg?

In [ ]:
top_popular_movies = movies.sort_values("popularity", ascending=False)
top_10_popular = top_popular_movies.head(10)
top_10_popular

id     budget  popularity     revenue  \
58   43692  165000000         724   675120017   
78   43724  150000000         434   378858340   
119  43796  140000000         271   655011224   
120  43797  125000000         206   752100229   
45   43662  185000000         187  1004558444   
162  43867  108000000         167   630161890   
59   43693  160000000         167   825532764   
5    43606  250000000         155   873260194   
0    43597  237000000         150  2787965087   
383  44259   63000000         146   100853753   

                                                 title  vote_average  \
58                                        Interstellar           8.1   
78                                  Mad Max: Fury Road           7.2   
119  Pirates of the Caribbean: The Curse of the Bla...           7.5   
120              The Hunger Games: Mockingjay - Part 1           6.6   
45                                     The Dark Knight           8.2   
162                                        The Martian           7.6   
59                                           Inception           8.1   
5                   Batman v Superman: Dawn of Justice           5.7   
0                                               Avatar           7.2   
383                                         Fight Club           8.3   

     vote_count  director_id  year month        day  
58        10867         4765  2014   Nov  Wednesday  
78         9427         4845  2015   May  Wednesday  
119        6985         4763  2003   Jul  Wednesday  
120        5584         4831  2014   Nov    Tuesday  
45        12002         4765  2008   Jul  Wednesday  
162        7268         4779  2015   Sep  Wednesday  
59        13752         4765  2010   Jul  Wednesday  
5          7004         4771  2016   Mar  Wednesday  
0         11800         4762  2009   Dec   Thursday  
383        9413         4829  1999   Oct     Friday

In [ ]:
top_voted = top_10_popular.sort_values("vote_average", ascending=False)
top_voted

id     budget  popularity     revenue  \
383  44259   63000000         146   100853753   
45   43662  185000000         187  1004558444   
58   43692  165000000         724   675120017   
59   43693  160000000         167   825532764   
162  43867  108000000         167   630161890   
119  43796  140000000         271   655011224   
78   43724  150000000         434   378858340   
0    43597  237000000         150  2787965087   
120  43797  125000000         206   752100229   
5    43606  250000000         155   873260194   

                                                 title  vote_average  \
383                                         Fight Club           8.3   
45                                     The Dark Knight           8.2   
58                                        Interstellar           8.1   
59                                           Inception           8.1   
162                                        The Martian           7.6   
119  Pirates of the Caribbean: The Curse of the Bla...           7.5   
78                                  Mad Max: Fury Road           7.2   
0                                               Avatar           7.2   
120              The Hunger Games: Mockingjay - Part 1           6.6   
5                   Batman v Superman: Dawn of Justice           5.7   

     vote_count  director_id  year month        day  
383        9413         4829  1999   Oct     Friday  
45        12002         4765  2008   Jul  Wednesday  
58        10867         4765  2014   Nov  Wednesday  
59        13752         4765  2010   Jul  Wednesday  
162        7268         4779  2015   Sep  Wednesday  
119        6985         4763  2003   Jul  Wednesday  
78         9427         4845  2015   May  Wednesday  
0         11800         4762  2009   Dec   Thursday  
120        5584         4831  2014   Nov    Tuesday  
5          7004         4771  2016   Mar  Wednesday

In [ ]:
ans = top_voted.iloc[[1]]
ans

id     budget  popularity     revenue            title  vote_average  \
45  43662  185000000         187  1004558444  The Dark Knight           8.2   

    vote_count  director_id  year month        day  
45       12002         4765  2008   Jul  Wednesday

---

### Quiz: Is the below code doing the same thing?

In [ ]:
top_popular_movies = movies.sort_values(["popularity", "vote_average"], ascending=[False, False]).head(10).iloc[[1]]
top_popular_movies

id     budget  popularity    revenue               title  vote_average  \
78  43724  150000000         434  378858340  Mad Max: Fury Road           7.2   

    vote_count  director_id  year month        day  
78        9427         4845  2015   May  Wednesday

---

### Q10 - How many movies that are among top 10 popular movies are also among top 10 rated (vote_avg)?

In [ ]:
# Get the top 10 popular movies

top_10_popular_movies = movies.sort_values("popularity", ascending=False)["title"].head(10)
top_10_popular_movies

58                                          Interstellar
78                                    Mad Max: Fury Road
119    Pirates of the Caribbean: The Curse of the Bla...
120                The Hunger Games: Mockingjay - Part 1
45                                       The Dark Knight
162                                          The Martian
59                                             Inception
5                     Batman v Superman: Dawn of Justice
0                                                 Avatar
383                                           Fight Club
Name: title, dtype: object

In [ ]:
# Get the top 10 rated movies

top_10_rated_movies = movies.sort_values("vote_average", ascending=False)["title"].head(10)
top_10_rated_movies

1284                                     Pulp Fiction
383                                        Fight Club
901                                  Schindler's List
459                                      Forrest Gump
45                                    The Dark Knight
917                                        GoodFellas
199     The Lord of the Rings: The Return of the King
785                                             Se7en
59                                          Inception
1326                               The Usual Suspects
Name: title, dtype: object

In [ ]:
df_popularly_rated = movies[(movies["title"].isin(top_10_popular_movies)) & (movies["title"].isin(top_10_rated_movies))]
df_popularly_rated

id     budget  popularity     revenue            title  vote_average  \
45   43662  185000000         187  1004558444  The Dark Knight           8.2   
59   43693  160000000         167   825532764        Inception           8.1   
383  44259   63000000         146   100853753       Fight Club           8.3   

     vote_count  director_id  year month        day  
45        12002         4765  2008   Jul  Wednesday  
59        13752         4765  2010   Jul  Wednesday  
383        9413         4829  1999   Oct     Friday

---

### Q11 - Create a new metric called `revenue_per_budget` i.e. revenue/budget and round it off up to 2 decimal places.

In [ ]:
def new_metric(df):
  val = np.round(df["revenue"]/df["budget"],2)
  df["revenue/budget"] = val
  return df

In [ ]:
movies_copy = movies[["revenue", "budget"]].apply(new_metric, axis=1)
movies_copy.head()

<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: divide by zero encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: invalid value encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: divide by zero encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: invalid value encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: invalid value encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: divide by zero encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: invalid value encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-i

revenue       budget  revenue/budget
0  2.787965e+09  237000000.0           11.76
1  9.610000e+08  300000000.0            3.20
2  8.806746e+08  245000000.0            3.59
3  1.084939e+09  250000000.0            4.34
4  8.908716e+08  258000000.0            3.45

In [ ]:
movies_copy = movies[["revenue", "budget"]].apply(new_metric, axis=1)
movies_copy.head()

<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: divide by zero encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: invalid value encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: divide by zero encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: invalid value encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: invalid value encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: divide by zero encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-input-48-60f48eaaf896>:2: RuntimeWarning: invalid value encountered in scalar divide
  val = np.round(df["revenue"]/df["budget"],2)
<ipython-i

revenue       budget  revenue/budget
0  2.787965e+09  237000000.0           11.76
1  9.610000e+08  300000000.0            3.20
2  8.806746e+08  245000000.0            3.59
3  1.084939e+09  250000000.0            4.34
4  8.908716e+08  258000000.0            3.45

---

### Q12 - In which year did Christopher Nolan produced most number of movies?

In [ ]:
data = movies.merge(directors, how='left', left_on='director_id', right_on='id')
data.drop(['director_id','id_y'], axis=1, inplace=True)

In [ ]:
chris_movies = data[data["director_name"] == "Christopher Nolan"]
print(chris_movies)

       id_x     budget  popularity     revenue                  title  \
3     43600  250000000         112  1084939099  The Dark Knight Rises   
45    43662  185000000         187  1004558444        The Dark Knight   
58    43692  165000000         724   675120017           Interstellar   
59    43693  160000000         167   825532764              Inception   
74    43716  150000000         115   374218673          Batman Begins   
565   44630   46000000          41   113714830               Insomnia   
641   44793   40000000          74   109676311           The Prestige   
1341  47170    9000000          60    39723096                Memento   

      vote_average  vote_count  year month        day      director_name  \
3              7.6        9106  2012   Jul     Monday  Christopher Nolan   
45             8.2       12002  2008   Jul  Wednesday  Christopher Nolan   
58             8.1       10867  2014   Nov  Wednesday  Christopher Nolan   
59             8.1       13752  2010  

In [ ]:
result = chris_movies.groupby("year")["title"].count()
result

year
2000    1
2002    1
2005    1
2006    1
2008    1
2010    1
2012    1
2014    1
Name: title, dtype: int64

Let's us write a function to do the exact same thing for whichever director we want.

In [ ]:
def year_wise_movies(director_name):

  dir_movies = data[data["director_name"] == director_name]

  result = dir_movies.groupby("year")["title"].count()

  return result

In [ ]:
year_wise_movies('James Cameron')

year
1984    1
1986    1
1989    1
1991    1
1994    1
1997    1
2009    1
Name: title, dtype: int64

In [ ]:
# grouping data on multiple columns

data.groupby(["director_name","year"])["title"].count()

director_name  year
Adam McKay     2004    1
               2006    1
               2008    1
               2010    1
               2013    1
                      ..
Zhang Yimou    2004    1
               2006    1
               2009    1
               2011    1
               2014    1
Name: title, Length: 1398, dtype: int64

In [ ]:
res = data.groupby(["director_name","year"])["title"].count()
res.sort_values(ascending = False)

director_name      year
Michael Moore      2007    2
Steven Soderbergh  2011    2
Robert Zemeckis    2000    2
John McTiernan     1999    2
Shawn Levy         2014    2
                          ..
James Mangold      1999    1
                   1997    1
James Ivory        2009    1
                   2005    1
Zhang Yimou        2014    1
Name: title, Length: 1398, dtype: int64

In [ ]:
data[["director_name", "year"]].value_counts() # one-liner logic

director_name      year
Michael Moore      2007    2
Steven Soderbergh  2011    2
Robert Zemeckis    2000    2
John McTiernan     1999    2
Shawn Levy         2014    2
                          ..
James Mangold      1999    1
                   1997    1
James Ivory        2009    1
                   2005    1
Zhang Yimou        2014    1
Length: 1398, dtype: int64

---

### Q13 - In which decade, the most number of movies were produced?

In [ ]:
data.describe()

id_x        budget   popularity       revenue  vote_average  \
count   1465.000000  1.465000e+03  1465.000000  1.465000e+03   1465.000000   
mean   45225.191126  4.802295e+07    30.855973  1.432539e+08      6.368191   
std     1189.096396  4.935541e+07    34.845214  2.064918e+08      0.818033   
min    43597.000000  0.000000e+00     0.000000  0.000000e+00      3.000000   
25%    44236.000000  1.400000e+07    11.000000  1.738013e+07      5.900000   
50%    45022.000000  3.300000e+07    23.000000  7.578164e+07      6.400000   
75%    45990.000000  6.600000e+07    41.000000  1.792469e+08      6.900000   
max    48395.000000  3.800000e+08   724.000000  2.787965e+09      8.300000   

         vote_count         year  
count   1465.000000  1465.000000  
mean    1146.396587  2002.615017  
std     1578.077438     8.680141  
min        1.000000  1976.000000  
25%      216.000000  1998.000000  
50%      571.000000  2004.000000  
75%     1387.000000  2009.000000  
max    13752.000000  2016.000000

To get an idea about the range of years we are dealing with.

We have data from 1976 to 2016 - let's create the decades accordingly.

In [ ]:
# Binning the years into decades -

bins = [1970, 1980, 1990, 2000, 2010, 2020]
labels = ["1", "2", "3", "4", "5"]

data["year_bucket"] = pd.cut(data["year"], bins = bins, labels = labels)
data

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name gender year_bucket  
0     2009   Dec  Thursday      James Cameron   Male           4  
1     2007   May  Saturday     Gore Verbinski   Male           4  
2     2015   Oct    Monday         Sam Mendes   Male           5  
3     2012   Jul    Monday  Christopher Nolan   Male           5  
4     2007   May   Tuesday          Sam Raimi   Male           4  
...    ...   ...       ...                ...    ...         ...  
1460  1978   May    Monday    Martin Scorsese   Male           1  
1461  1994   Sep   Tuesday        Kevin Smith   Male           3  
1462  2009   Aug    Friday           Uwe Boll   Male           4  
1463  1990   Jul    Friday  Richard Linklater   Male           2  
1464  1992   Sep    Friday   Robert Rodriguez    NaN           3  

[1465 rows x 13 columns]

In [ ]:
data["year_bucket"].value_counts()

4    644
3    376
5    293
2    126
1     26
Name: year_bucket, dtype: int64

Most number of movies were produced in between the years 2000 and 2010.

---

### Q14 - Within the decade with most number of movies, which were the top 5 contributing directors (who produced most)?

In [ ]:
data_4 = data[data["year_bucket"] == 4]
data["director_name"].value_counts().head()

Steven Spielberg    26
Martin Scorsese     19
Clint Eastwood      19
Woody Allen         18
Ridley Scott        16
Name: director_name, dtype: int64

In [ ]:
# Check for missing values
data.isna().sum()

id_x               0
budget             0
popularity         0
revenue            0
title              0
vote_average       0
vote_count         0
year               0
month              0
day                0
director_name      0
gender           124
year_bucket        0
dtype: int64

In [ ]:
# Filling null values
data["gender"].fillna("NA", inplace = True)

In [ ]:
data["gender"].value_counts(normalize=True)*100

Male      89.351536
NA         8.464164
Female     2.184300
Name: gender, dtype: float64

### HW - Try to fill gender with M/F depending on the mode within each genre.

---

### Q15 - Display movie details of only those years where number of movies produced are more than 70.

In [ ]:
result = data.groupby("year").filter(lambda x: x["title"].count() > 70)
result

id_x     budget  popularity     revenue  \
6     43607  270000000          57   391081192   
8     43609  200000000         145  1065659812   
18    43621  207000000          61   550000000   
24    43630  210000000           3   459359555   
43    43660  180000000          67   748806957   
...     ...        ...         ...         ...   
1342  47172    5000000          18    26888376   
1359  47275          0           3           0   
1365  47328    4200000          16           0   
1373  47351    4000000          17    31584901   
1407  47696    2000000          12     5967038   

                                                  title  vote_average  \
6                                      Superman Returns           5.4   
8            Pirates of the Caribbean: Dead Man's Chest           7.0   
18                                            King Kong           6.6   
24                                X-Men: The Last Stand           6.3   
43    The Chronicles of Narnia: The Lion, the Witch ...           6.7   
...                                                 ...           ...   
1342                                          Clerks II           6.9   
1359                              A Cock and Bull Story           6.2   
1365                                High School Musical           6.1   
1373                                              Scoop           6.4   
1407                                        Harsh Times           6.3   

      vote_count  year month        day         director_name gender  \
6           1400  2006   Jun  Wednesday          Bryan Singer   Male   
8           5246  2006   Jun    Tuesday        Gore Verbinski   Male   
18          2337  2005   Dec  Wednesday         Peter Jackson   Male   
24          3525  2006   May  Wednesday          Brett Ratner   Male   
43          2629  2005   Dec  Wednesday        Andrew Adamson   Male   
...          ...   ...   ...        ...                   ...    ...   
1342         398  2006   May   Thursday           Kevin Smith   Male   
1359          43  2005   Jul     Sunday  Michael Winterbottom   Male   
1365        1000  2006   Jan     Friday          Kenny Ortega   Male   
1373         420  2006   Jul   Thursday           Woody Allen   Male   
1407         197  2005   Sep     Sunday            David Ayer     NA   

     year_bucket  
6              4  
8              4  
18             4  
24             4  
43             4  
...          ...  
1342           4  
1359           4  
1365           4  
1373           4  
1407           4  

[144 rows x 13 columns]

In [ ]:
data["year"].value_counts()

2006    73
2005    71
2002    69
2008    67
2011    67
2009    66
1999    64
2010    63
2001    62
2004    61
2000    59
2003    58
2013    57
2007    54
2014    53
2012    49
1997    46
1996    46
1998    45
2015    41
1995    33
1994    31
2016    26
1992    22
1989    18
1990    18
1993    18
1985    14
1988    13
1991    12
1987    12
1986    11
1981    11
1982    10
1984    10
1983     9
1980     8
1977     6
1978     5
1979     4
1976     3
Name: year, dtype: int64

Only year 2005 and 2006 had more than 70 movies, i.e. 71 and 73 respectively so total rows in the output should be 71+73 = 144

In [ ]:
data[data["year"].isin([2005, 2006])]

id_x     budget  popularity     revenue  \
6     43607  270000000          57   391081192   
8     43609  200000000         145  1065659812   
18    43621  207000000          61   550000000   
24    43630  210000000           3   459359555   
43    43660  180000000          67   748806957   
...     ...        ...         ...         ...   
1342  47172    5000000          18    26888376   
1359  47275          0           3           0   
1365  47328    4200000          16           0   
1373  47351    4000000          17    31584901   
1407  47696    2000000          12     5967038   

                                                  title  vote_average  \
6                                      Superman Returns           5.4   
8            Pirates of the Caribbean: Dead Man's Chest           7.0   
18                                            King Kong           6.6   
24                                X-Men: The Last Stand           6.3   
43    The Chronicles of Narnia: The Lion, the Witch ...           6.7   
...                                                 ...           ...   
1342                                          Clerks II           6.9   
1359                              A Cock and Bull Story           6.2   
1365                                High School Musical           6.1   
1373                                              Scoop           6.4   
1407                                        Harsh Times           6.3   

      vote_count  year month        day         director_name gender  \
6           1400  2006   Jun  Wednesday          Bryan Singer   Male   
8           5246  2006   Jun    Tuesday        Gore Verbinski   Male   
18          2337  2005   Dec  Wednesday         Peter Jackson   Male   
24          3525  2006   May  Wednesday          Brett Ratner   Male   
43          2629  2005   Dec  Wednesday        Andrew Adamson   Male   
...          ...   ...   ...        ...                   ...    ...   
1342         398  2006   May   Thursday           Kevin Smith   Male   
1359          43  2005   Jul     Sunday  Michael Winterbottom   Male   
1365        1000  2006   Jan     Friday          Kenny Ortega   Male   
1373         420  2006   Jul   Thursday           Woody Allen   Male   
1407         197  2005   Sep     Sunday            David Ayer     NA   

     year_bucket  
6              4  
8              4  
18             4  
24             4  
43             4  
...          ...  
1342           4  
1359           4  
1365           4  
1373           4  
1407           4  

[144 rows x 13 columns]

---